<!DOCTYPE html>
<html lang="en">
  <head>
    <meta http-equiv="X-UA-Compatible" content="IE=edge">
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <meta charset="utf-8">
    <title>MpileupToVcf</title>

    <!-- Latest compiled and minified CSS -->
    <link rel="stylesheet" href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">

    <!-- Optional theme -->
    <link rel="stylesheet" href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap-theme.min.css" integrity="sha384-rHyoN1iRsVXV4nD0JutlnGaslCJuC7uwjduW9SVrLvRYooPp2bWYgmgJQIXwl/Sp" crossorigin="anonymous">

    <!-- jQuery (necessary for Bootstrap's JavaScript plugins) -->
    <script src="https://ajax.googleapis.com/ajax/libs/jquery/1.12.4/jquery.min.js"></script>

    <!-- Latest compiled and minified JavaScript -->
    <script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script>

    <!--[if lt IE 9]>
      <script src="http://html5shim.googlecode.com/svn/trunk/html5.js"></script>
    <![endif]-->
    <link href="css/docs.css" rel="stylesheet">

  </head>

  <body>
  <nav class="navbar navbar-default navbar-fixed-top">
    <div class="container-fluid">
      <div class="navbar-header">
        <div class="navbar-header">
          <a class="navbar-brand" href="#">MpileupToVcf</a>
          <ul class="nav navbar-nav navbar-right">
            <li class="navbar-text" id="Version">
              Version</li>
          </ul>
        </div>
      </div>



    </div>

  </nav>
  <div class="container">

    <div class="row">
      <div class="col-sm-3 col-md-2 sidebar" >

        <ul class="nav nav-sidebar nav-list">
          <li><b>Index</b></li>
          
          <li><a href="#description"><span>Description</span></a></li>
          
          <li><a href="#installation"><span>Installation</span></a></li>
          
          <li><a href="#manual"><span>Manual</span></a></li>
          
          <li><a href="#about"><span>About</span></a></li>
          
          <li><a href="#contributing"><span>Contributing</span></a></li>
          
          <li><a href="#contact"><span>Contact</span></a></li>
          
        </ul>
        <ul class="nav nav-sidebar nav-list">
          <li><b>Project Links</b></li>

          <li><a href="https://github.com/biopet/mpileuptovcf">Source code</a></li>
          <li><a href="https://github.com/biopet/mpileuptovcf/issues">Feedback, feature requests and bug reports</a></li>
          <li><a href="api/index.html">API Documentation</a></li>
        </ul>
        <ul class="nav nav-sidebar nav-list">
          <li><b>Documentation</b></li>
          <li><a href="../index.html">Latest stable version</a></li>
          <li><a href="../develop/index.html">Development version</a></li>
        </ul>
      </div>
      
      <div class="col-sm-9 col-sm-offset-3 col-md-10 col-md-offset-2 main" id="top">

        
        <h1 id="description" class="section">Description</h1>
        <p>This tool enables a user to extract a VCF file out a mpileup file generated from the BAM file using <em>samtools mpileup</em>,
        for instance. The tool can also stream through STDin so that it is not necessary to store the mpileup file
        on disk. Mpileup files can to be very large because they describe each covered base position in the genome on a per
        read basis, so it is not desired to store them.</p>
        
        <h1 id="installation" class="section">Installation</h1>
        <p>MpileupToVcf requires Java 8 to be installed on your device. Download Java 8
        <a href="http://www.oracle.com/technetwork/java/javase/downloads/jre8-downloads-2133155.html">here</a>
        or install via your distribution&#39;s package manager.</p>
        <p>Download the latest version of MpileupToVcf <a href="https://github.com/biopet/mpileuptovcf/releases/">here</a>.
        To generate the usage run:</p>
        <pre><code>java -jar &lt;MpileupToVcf_jar&gt; --help</code></pre>
        
        <h1 id="manual" class="section">Manual</h1>
        <p>MpileupToVcf comes with various options. See the usage for more details. The tool can stream from stdin or accept
        a mpileup file. An output file and the name of the sample are always required.</p>
        
        <h2 id="example" class="section">Example</h2>
        <p>To convert a mpileup file to vcf from a haploid organism and an expected sequencing error rate of 0.010&quot;</p>
        <pre><code>java -jar &lt;MpileupToVcf_jar&gt; \
-I input.mpileup \
-o output.vcf \
--sample Yeast5302 \
--ploidy 1 \
--seqError 0.010</code></pre>
        <p>To convert a mpileup directly from standard out:</p>
        <pre><code>samtools mpileup &lt;bam&gt; | java -jar &lt;MpileupToVcf_jar&gt; \
-o &lt;output_vcf&gt; \
--sample E.coli243</code></pre>
        
        <h2 id="usage" class="section">Usage</h2>
        <p>Usage for MpileupToVcf:</p>
        <p>&lt;table class=&quot;table&quot;&gt;
          &lt;thead&gt;
            &lt;tr&gt;
              &lt;th&gt;Option&lt;/th&gt;
              &lt;th&gt;Required&lt;/th&gt;
              &lt;th&gt;Can occur multiple times&lt;/th&gt;
              &lt;th&gt;Description&lt;/th&gt;
            &lt;/tr&gt;
          &lt;/thead&gt;
          &lt;tbody&gt;
            &lt;tr&gt;
              &lt;td&gt;--log_level, -l&lt;/td&gt;
              &lt;td&gt;no&lt;/td&gt;
              &lt;td&gt;no&lt;/td&gt;
              &lt;td&gt;Level of log information printed. Possible levels: &#39;debug&#39;, &#39;info&#39;, &#39;warn&#39;, &#39;error&#39;&lt;/td&gt;
            &lt;/tr&gt;
            &lt;tr&gt;
              &lt;td&gt;--help, -h&lt;/td&gt;
              &lt;td&gt;no&lt;/td&gt;
              &lt;td&gt;no&lt;/td&gt;
              &lt;td&gt;Print usage&lt;/td&gt;
            &lt;/tr&gt;
            &lt;tr&gt;
              &lt;td&gt;--version, -v&lt;/td&gt;
              &lt;td&gt;no&lt;/td&gt;
              &lt;td&gt;no&lt;/td&gt;
              &lt;td&gt;Print version&lt;/td&gt;
            &lt;/tr&gt;
            &lt;tr&gt;
              &lt;td&gt;--input, -I&lt;/td&gt;
              &lt;td&gt;no&lt;/td&gt;
              &lt;td&gt;no&lt;/td&gt;
              &lt;td&gt;input, default is stdin&lt;/td&gt;
            &lt;/tr&gt;
            &lt;tr&gt;
              &lt;td&gt;--output, -o&lt;/td&gt;
              &lt;td&gt;yes&lt;/td&gt;
              &lt;td&gt;no&lt;/td&gt;
              &lt;td&gt;output file (required)&lt;/td&gt;
            &lt;/tr&gt;
            &lt;tr&gt;
              &lt;td&gt;--sample, -s&lt;/td&gt;
              &lt;td&gt;yes&lt;/td&gt;
              &lt;td&gt;no&lt;/td&gt;
              &lt;td&gt;Sample name in the vcf file&lt;/td&gt;
            &lt;/tr&gt;
            &lt;tr&gt;
              &lt;td&gt;--minDP&lt;/td&gt;
              &lt;td&gt;no&lt;/td&gt;
              &lt;td&gt;no&lt;/td&gt;
              &lt;td&gt;Minimal total depth&lt;/td&gt;
            &lt;/tr&gt;
            &lt;tr&gt;
              &lt;td&gt;--minAP&lt;/td&gt;
              &lt;td&gt;no&lt;/td&gt;
              &lt;td&gt;no&lt;/td&gt;
              &lt;td&gt;Minimal alternative depth&lt;/td&gt;
            &lt;/tr&gt;
            &lt;tr&gt;
              &lt;td&gt;--homoFraction&lt;/td&gt;
              &lt;td&gt;no&lt;/td&gt;
              &lt;td&gt;no&lt;/td&gt;
              &lt;td&gt;If alleles are above this fraction it&#39;s being seen as homozygous. Default if 0.8&lt;/td&gt;
            &lt;/tr&gt;
            &lt;tr&gt;
              &lt;td&gt;--ploidy&lt;/td&gt;
              &lt;td&gt;no&lt;/td&gt;
              &lt;td&gt;no&lt;/td&gt;
              &lt;td&gt;Specify the ploidy as a number: &#39;1&#39; for haploid, &#39;2&#39; for diploid etc.&lt;/td&gt;
            &lt;/tr&gt;
            &lt;tr&gt;
              &lt;td&gt;--seqError&lt;/td&gt;
              &lt;td&gt;no&lt;/td&gt;
              &lt;td&gt;no&lt;/td&gt;
              &lt;td&gt;Expected sequencing error rate, default is 0.005&lt;/td&gt;
            &lt;/tr&gt;
            &lt;tr&gt;
              &lt;td&gt;--refCalls&lt;/td&gt;
              &lt;td&gt;no&lt;/td&gt;
              &lt;td&gt;no&lt;/td&gt;
              &lt;td&gt;If set refcalls are also writen. Warning: This will results in a very large vcf file&lt;/td&gt;
            &lt;/tr&gt;
          &lt;/tbody&gt;
        &lt;/table&gt;</p>
        
        <h1 id="about" class="section">About</h1>
        <p>MpileupToVcf is part of BIOPET tool suite that is developed at LUMC by <a href="http://sasc.lumc.nl/">the SASC team</a>.
        Each tool in the BIOPET tool suite is meant to offer a standalone function that can be used to perform a
        dedicate data analysis task or added as part of <a href="http://biopet-docs.readthedocs.io/en/latest/">BIOPET pipelines</a>.</p>
        <p>All tools in the BIOPET tool suite are <a href="https://www.gnu.org/philosophy/free-sw.html">Free/Libre</a> and
        <a href="https://opensource.org/osd">Open Source</a> Software.</p>
        
        <h1 id="contributing" class="section">Contributing</h1>
        <p>The source code of MpileupToVcf can be found <a href="https://github.com/biopet/mpileuptovcf">here</a>.
        We welcome any contributions. Bug reports, feature requests and feedback can be submitted at our
        <a href="https://github.com/biopet/mpileuptovcf/issues">issue tracker</a>.</p>
        <p>MpileupToVcf is build using sbt. Before submitting a pull request, make sure all tests can be passed by
        running <code>sbt test</code> from the project&#39;s root. We recommend using an IDE to work on MpileupToVcf. We have had
        good results with <a href="https://www.jetbrains.com/idea/">this IDE</a>.</p>
        
        <h1 id="contact" class="section">Contact</h1>
        <p>&lt;p&gt;
          &lt;!-- Obscure e-mail address for spammers --&gt;
        For any question related to MpileupToVcf, please use the
        &lt;a href=&#39;https://github.com/biopet/mpileuptovcf/issues&#39;&gt;github issue tracker&lt;/a&gt;
        or contact
         &lt;a href=&#39;http://sasc.lumc.nl/&#39;&gt;the SASC team&lt;/a&gt; directly at: &lt;a href=&#39;&amp;#109;&amp;#97;&amp;#105;&amp;#108;&amp;#116;&amp;#111;&amp;#58;&amp;#115;&amp;#97;&amp;#115;&amp;#99;&amp;#64;&amp;#108;&amp;#117;&amp;#109;&amp;#99;&amp;#46;&amp;#110;&amp;#108;&#39;&gt;
        &amp;#115;&amp;#97;&amp;#115;&amp;#99;&amp;#64;&amp;#108;&amp;#117;&amp;#109;&amp;#99;&amp;#46;&amp;#110;&amp;#108;&lt;/a&gt;.
        &lt;/p&gt;</p>      

      </div>
    </div>

  </div>
  <script>
    var loc = window.location.pathname;
    var thisDir = loc.substring(0, loc.lastIndexOf('/'));
    var splitLocation = window.location.pathname.split('/');
    var version = splitLocation[splitLocation.length - 2]
    document.getElementById("Version").innerHTML = "Version " + version;
  </script>
</body>

</html>
